core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectory.java

```diff
@@ -39,16 +41,15 @@
  */
 @NotThreadSafe
 public final class InodeDirectory extends Inode<InodeDirectory> implements InodeDirectoryView {
-  private static final IndexDefinition<Inode<?>, String> NAME_INDEX =
-      new IndexDefinition<Inode<?>, String>(true) {
-        @Override
-        public String getFieldValue(Inode<?> o) {
-          return o.getName();
-        }
-      };
+  // Profiling shows that most of the file lists are between 1 and 4 elements.
+  // Thus allocate the corresponding ArrayLists with a small initial capacity.
+  private static final int DEFAULT_FILES_PER_DIRECTORY = 2;
+
+  private int searchChildren(String name) {
+    return mChildren == null ? -1 : Collections.binarySearch(mChildren, name);
+  }
 
-  /** Use UniqueFieldIndex directly for name index rather than using IndexedSet. */
-  private final FieldIndex<Inode<?>, String> mChildren = new UniqueFieldIndex<>(NAME_INDEX);
+  private List<Inode<?>> mChildren = null;
```

Comment:
Will the performance be much worse for directories with large number of children? If so, could we abstract the data structure and migrate behind the scenes between the list and map?

Reply:
This PR requires more performance testing before make a judgment.